node/overseer/overseer-gen/src/lib.rs

```diff
@@ -0,0 +1,519 @@
+		}
+	}
+
+}
+
+
+
+/// A helper trait to map a subsystem to smth. else.
+pub trait MapSubsystem<T> {
+	/// The output type of the mapping.
+	type Output;
+
+	/// Consumes a `T` per subsystem, and maps it to `Self::Output`.
+	fn map_subsystem(&self, sub: T) -> Self::Output;
+}
+
+impl<F, T, U> MapSubsystem<T> for F where F: Fn(T) -> U {
+	type Output = U;
+
+	fn map_subsystem(&self, sub: T) -> U {
+		(self)(sub)
+	}
+}
+
+/// A wrapping type for messages.
+///
+/// Includes a counter to synchronize signals with messages,
+/// such that no inconsistent message sequences are prevented.
+#[derive(Debug)]
+pub struct MessagePacket<T> {
+	/// Signal level at the point of reception.
+	///
+	/// Required to assure signals were consumed _before_
+	/// consuming messages that are based on the assumption
+	/// that a certain signal was assumed.
+	pub signals_received: usize,
+	/// The message to be sent/consumed.
+	pub message: T,
+}
+
+/// Create a packet from its parts.
+pub fn make_packet<T>(signals_received: usize, message: T) -> MessagePacket<T> {
+	MessagePacket {
+		signals_received,
+		message,
+	}
+}
+
+/// Incoming messages from both the bounded and unbounded channel.
+pub type SubsystemIncomingMessages<M> = self::stream::Select<
+	self::metered::MeteredReceiver<MessagePacket<M>>,
+	self::metered::UnboundedMeteredReceiver<MessagePacket<M>>,
+>;
+
+
+/// Watermark to track the received signals.
+#[derive(Debug, Default, Clone)]
+pub struct SignalsReceived(Arc<AtomicUsize>);
+
+impl SignalsReceived {
+	/// Load the current value of received signals.
+	pub fn load(&self) -> usize {
+		// off by a few is ok
+		self.0.load(atomic::Ordering::Relaxed)
+	}
+
+	/// Increase the number of signals by one.
+	pub fn inc(&self) {
+		self.0.fetch_add(1, atomic::Ordering::Acquire);
+	}
+}
+
+
+
+/// A trait to support the origin annotation
+/// such that errors across subsystems can be easier tracked.
+pub trait AnnotateErrorOrigin: 'static + Send + Sync + std::error::Error {
+	/// Annotate the error with a origin `str`.
+	///
+	/// Commonly this is used to create nested enum variants.
+	///
+	/// ```rust,ignore
+	/// E::WithOrigin("I am originally from Cowtown.", E::Variant)
+	/// ```
+	fn with_origin(self, origin: &'static str) -> Self;
+}
+
+/// An asynchronous subsystem task..
+///
+/// In essence it's just a newtype wrapping a `BoxFuture`.
+pub struct SpawnedSubsystem<E>
+	where
+		E: std::error::Error
+			+ Send
+			+ Sync
+			+ 'static
+			+ From<self::OverseerError>,
+{
+	/// Name of the subsystem being spawned.
+	pub name: &'static str,
+	/// The task of the subsystem being spawned.
+	pub future: BoxFuture<'static, Result<(), E>>,
+}
+
+/// An error type that describes faults that may happen
+///
+/// These are:
+///   * Channels being closed
+///   * Subsystems dying when they are not expected to
+///   * Subsystems not dying when they are told to die
+///   * etc.
+// FIXME XXX make generic over the source error of FromOrigin
```

Comment:
is this tracked in #3427? what would be the benefit of that change?

Reply:
That particular one is obsolete, made sure to add references to all TODO tags in `node/overseer` and have a bullet point in #3427 chores section